### PR TITLE
Docker: slim down image size, by removing temp/cache files

### PR DESCRIPTION
**Pull request by TheCleric**
Per the [note](https://github.com/nativefier/nativefier/pull/1122#discussion_r588922780) by @SuperSandro2000 in #1122 Docker will still cache files in intermediate layers if you delete them, so they'll still be part of the image.

Only solution seems to be to delete them as you create them so they don't cache:
Per https://stackoverflow.com/questions/53998310/docker-remove-file-from-intermediate-layer

### CODE REVIEW

**Comment by SuperSandro2000** [comment]
Alternative is to squash the layers after they are build.

File: Dockerfile [hunk 1]
```diff
@@ -21,25 +21,24 @@ RUN npm install
 WORKDIR /nativefier
 
 # Install (note that we had to manually install in `app` before, as `prepare` won't run as root)
-# Also, running tests, to ensure we don't Docker build & publish broken stuff
-RUN npm install && npm run build && npm test && npm link
+# Also, running tests, to ensure we don't Docker build & publish broken stuff, and cleanup test files
+RUN npm install && npm run build && npm test && npm link && rm -rf /tmp/nativefier*
```
Comment (SuperSandro2000) [inline Dockerfile:25]:
```suggestion
RUN npm install \
  && npm run build \
  && npm test \
  && npm link \
  && rm -rf /tmp/nativefier*
```

**Review by SuperSandro2000 — APPROVED** [review]
looks good but you can check it with https://github.com/wagoodman/dive to be sure.

**Comment by TheCleric** [comment]
> looks good but you can check it with https://github.com/wagoodman/dive to be sure.

Wow, that tool is fantastic. Found a few more tweaks to help keep layer sizes down. Thanks!

**Comment by TheCleric** [comment]
@ronjouch based on the suggestions @SuperSandro2000 provided here, along with using his suggestion of checking it with Dive, I was able to get the generated image (including files "deleted" but still taking up space in intermediate layers) down from 2.6 GB for current `master` down to 1.0 GB in this PR.

**Comment by SuperSandro2000** [comment]
> Wow, that tool is fantastic. Found a few more tweaks to help keep layer sizes down. Thanks!

Glad I could help you.

File: Dockerfile [hunk 2]
```diff
@@ -12,34 +12,41 @@ WORKDIR /nativefier
 # Add sources
 COPY . .
```
[the remark this answers is not in the document]
Comment (TheCleric) [reply Dockerfile:13]: Doesn't seem to work ☹️ 

```
#12 21.75 npm ERR! code EACCES
#12 21.75 npm ERR! syscall access
#12 21.75 npm ERR! path /nativefier
#12 21.75 npm ERR! errno -13
#12 21.76 npm ERR! Error: EACCES: permission denied, access '/nativefier'
#12 21.76 npm ERR!  [Error: EACCES: permission denied, access '/nativefier'] {
#12 21.76 npm ERR!   errno: -13,
#12 21.76 npm ERR!   code: 'EACCES',
#12 21.76 npm ERR!   syscall: 'access',
#12 21.76 npm ERR!   path: '/nativefier'
#12 21.76 npm ERR! }
```

[the remark this answers is not in the document]
Comment (TheCleric) [reply Dockerfile:13]: Doing an `ls` after that step reveals the real problem: node has perms to all the files, but not the dir itself to create folders for the build.

File: Dockerfile [hunk 3]
```diff
@@ -12,34 +12,41 @@ WORKDIR /nativefier
...
 
+# Give everything to node
```
Comment (ronjouch) [inline Dockerfile:15]: In comments, please document the "why", not just the "what". The "what" alone doesn't need much commenting when the code is clear/easy like it is here.

Here's a short talk I like a lot on the subject. It's about Git commit messages, but applies well to comments too. [PyCon 2014 / Greg Ward - Documenting history, or how to write great commit messages: say *why*, not what](https://www.youtube.com/watch?v=Jb6ij4eRu6c#t=378)

File: Dockerfile [hunk 4]
```diff
@@ -12,34 +12,41 @@ WORKDIR /nativefier
...
 
-# Cleanup test artifacts
-RUN rm -rf /tmp/nativefier*
+ENV NPM_PACKAGES="/home/node/.npm-packages"
+ENV PATH="$PATH:$NPM_PACKAGES/bin"
+ENV MANPATH="$MANPATH:$NPM_PACKAGES/share/man"
 
-# Use 1000 as default user not root
-USER 1000
+# Link (which will install and buld), run tests (to ensure we don't Docker build & publish broken stuff), and cleanup files
```
Comment (ronjouch) [inline Dockerfile:33]:
```suggestion
# Link (which will install and build), run tests (to ensure we don't Docker build & publish broken stuff), and cleanup files
```

File: Dockerfile [hunk 5]
```diff
@@ -12,34 +12,41 @@ WORKDIR /nativefier
...
+RUN npm link \
+    && npm test \
+    && rm -rf /tmp/nativefier* ~/.npm/_cacache ~/.cache/electron \
+    && chmod +x $NPM_PACKAGES/bin/nativefier
 
 # Run a {lin,mac,win} build
 # 1. to check installation was sucessful
 # 2. to cache electron distributables and avoid downloads at runtime
+# Also delete generated apps so they don't get added to the Docker layer
```
Comment (ronjouch) [inline Dockerfile:42]:
```suggestion
# Also delete generated apps so they don't get added to the Docker layer
# !Important! The `rm -rf` command must be in the same `RUN` command (using an `&&`), to not waste Docker layer space
```

File: Dockerfile [hunk 6]
```diff
@@ -12,34 +12,41 @@ WORKDIR /nativefier
...
+ENV NPM_PACKAGES="/home/node/.npm-packages"
+ENV PATH="$PATH:$NPM_PACKAGES/bin"
+ENV MANPATH="$MANPATH:$NPM_PACKAGES/share/man"
 
-# Use 1000 as default user not root
-USER 1000
+# Link (which will install and buld), run tests (to ensure we don't Docker build & publish broken stuff), and cleanup files
+RUN npm link \
+    && npm test \
+    && rm -rf /tmp/nativefier* ~/.npm/_cacache ~/.cache/electron \
```
Comment (ronjouch) [inline Dockerfile:36]: Hmmm why removing `~/.cache/electron` at this point?

File: Dockerfile [hunk 7]
```diff
@@ -12,34 +12,41 @@ WORKDIR /nativefier
...
-# Build nativefier and link globally
-WORKDIR /nativefier/app
-RUN npm install
-WORKDIR /nativefier
-
-# Install (note that we had to manually install in `app` before, as `prepare` won't run as root)
-# Also, running tests, to ensure we don't Docker build & publish broken stuff
-RUN npm install && npm run build && npm test && npm link
+# Setup a global packages location for "node" user so we can npm link
+RUN mkdir ~/.npm-packages \
```
Comment (ronjouch) [inline Dockerfile:26]: In a normal user machine, a hidden dir makes total sense (that's what I do on my machine too), but in a Docker, there's no point to it.

```suggestion
RUN mkdir ~/npm-packages \
```

(... plus needs edit in the following lines, but GitHub won't let me)

**Review by ronjouch — CHANGES_REQUESTED** [review]
@TheCleric a few nits. Then I'm not sure if the `COPY` discussion is addressed; tell me when it is and the nits are addressed, and I'll merge. Thanks for taking care about this! 🙂

**Comment by TheCleric** [comment]
> @TheCleric a few nits. Then I'm not sure if the `COPY` discussion is addressed; tell me when it is and the nits are addressed, and I'll merge. Thanks for taking care about this! 🙂

I found a way to make the COPY work and addressed your comments here 😄

**Review by ronjouch — APPROVED** [review]
Thaaaaaaaaaaaanks @TheCleric . Merging.

**Pull request merged**